src/aviary/functional.py

```diff
@@ -0,0 +1,166 @@
+        update_wrapper(self, start_fn)
+
+    def __call__(self, *args, **kwargs):
+        """Create a new environment instance."""
+        return FunctionalEnvironment(self.start_fn, self.tools.copy(), *args, **kwargs)
+
+    def tool(self, **tool_kwargs):
+        """Decorator to add tools to the environment."""
+
+        def decorator(func: Callable):
+            tool = Tool.from_function(
+                func, **tool_kwargs, allow_empty_param_descriptions=True
+            )
+            # Check for name conflicts
+            if any(t.info.name == tool.info.name for t in self.tools):
+                raise RuntimeError(f"Tool with name '{tool.info.name}' already exists")
+            self.tools.append(tool)
+            return func
+
+        return decorator
+
+
+class fenv:  # noqa: N801
+    """Factory class for creating functional environments."""
+
+    @staticmethod
+    def start():
+        """Initialize a new functional environment definition.
+
+        This decorator marks the starting point for defining a functional environment.
+        It should be applied to a function that sets up the initial state and first
+        observation for the environment. The decorated function will serve as the
+        base for adding tools via the .tool() decorator.
+
+        The decorated function should accept any desired parameters that configure
+        the environment instance. It must return a tuple containing:
+        1. The initial observation message (str)
+        2. A dict of any state variables to persist between steps
+
+        Example:
+            @fenv.start()
+            def my_env(topic: str):
+                return f"Write a story about {topic}", {"topic": topic}
+
+            @my_env.tool()
+            def print_story(story: str, state) -> None:
+                '''Print the story and complete the task'''
+                print(story)
+                state.reward = 1
+                state.done = True
+
+            # Usage
+            env = my_env(topic="space")
+            obs, tools = await env.reset()
+
+        Returns:
+            callable: A decorator function that transforms the decorated function into
+            an environment builder. The returned environment builder provides a .tool()
+            decorator for adding tools to the environment.
+
+        Notes:
+            - The state dict returned by the decorated function will automatically include
+            'reward' (float) and 'done' (bool) fields that can be modified by tools
+            - Tools can access the state dict via an optional 'state' parameter
+            - The environment implements the standard Gymnasium interface with async
+            reset() and step() methods
+        """
+
+        def decorator(func: Callable) -> EnvironmentBuilder:
+            if not callable(func):
+                raise TypeError("Decorator must be applied to a function")
+            return EnvironmentBuilder(func)
```

Comment:
Can we get `functools.wraps` in here

Reply:
That is treated with `update_wrapper` in `EnvironmentBuilder`